xarray/conventions/netcdftimeindex.py

```diff
@@ -120,23 +118,31 @@ def get_date_type(self):
     return type(self._data[0])
 
 
-def assert_all_same_netcdftime_datetimes(data):
-    from netcdftime._netcdftime import datetime
+def assert_all_valid_date_type(data):
+    from netcdftime import (
+        DatetimeJulian, DatetimeNoLeap, DatetimeAllLeap,
+        DatetimeGregorian, DatetimeProlepticGregorian, Datetime360Day)
 
-    if not isinstance(data[0], datetime):
+    valid_types = (DatetimeJulian, DatetimeNoLeap, DatetimeAllLeap,
+                   DatetimeGregorian, DatetimeProlepticGregorian,
+                   Datetime360Day)
+
+    sample = data[0]
+    date_type = type(sample)
+    if not isinstance(sample, valid_types):
         raise TypeError(
-            'NetCDFTimeIndex requires netcdftime._netcdftime.datetime'
-            ' objects.')
-    if not all(isinstance(value, type(data[0])) for value in data):
+            'NetCDFTimeIndex requires netcdftime._netcdftime.datetime '
```

Comment:
I still prefer to use `netcdftime.datetime`, since it's equivalent and doesn't refer to a private submodule (the same reason we don't reference `xarray.core.dataset.Dataset`). The repr for netcdftime datetime should probably be fixed to refer to the shorter path.